11_As an administrator, I need to edit an announcement

Jean-Beru opened the pull request:
Issue #8

romqin commented on src/SensioLabs/JobBoardBundle/Controller/BackendController.php:
```diff
@@ -19,11 +21,32 @@ public function listAction(Request $request)
-     * @Route("/backend/edit", name="backend_edit")
-     * @Template()
+     * @Route("/backend/announcements/{id}/edit", name="backend_edit")
+     * @ParamConverter("job")
```
> If you're using Doctrine, then you can optionally use the ParamConverter to automatically query for an entity and pass it as an argument to your controller.
[source](http://symfony.com/doc/current/best_practices/controllers.html#using-the-paramconverter)

You can remove this line.

romqin commented on src/SensioLabs/JobBoardBundle/Controller/BackendController.php:
```diff
@@ -19,11 +21,32 @@ public function listAction(Request $request)
+
+        if ($form->handleRequest($request)->isValid()) {
+            $em = $this->getDoctrine()->getManager();
+            $em->persist($job);
```
`persist` is required only when your object is not managed by the doctrine's UnitOfWork.
Here `$job` is fetched from the database so it's already managed by doctrine.

You can remove this line.

mmarchois commented on src/SensioLabs/JobBoardBundle/Form/JobType.php:
```diff
@@ -37,12 +39,32 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                     ->create('howToApply', 'text', ['required' => false])
                     ->addModelTransformer($this->purifierTransformer)
             );
+        if ($options['is_admin']) {
```
If you have time, you can improve this feature by using the form TypeExtension. Alexandre Salomé written a PDF about this very useful feature. See http://bit.ly/1phQBUE pages 42 to 50

romqin commented on src/SensioLabs/JobBoardBundle/Entity/User.php:
```diff
@@ -173,4 +182,24 @@ public function getSalt()
+    /**
+     * @param string $email
+     *
+     * @return User
```
`return $this`

jeremyFreeAgent commented on src/SensioLabs/JobBoardBundle/Controller/BackendController.php:
```diff
@@ -19,11 +21,32 @@ public function listAction(Request $request)
+            $em->persist($job);
+            $em->flush();
+
+            if ($job->getValidated()) {
```
`getValidated` should be `isValidated`.

jeremyFreeAgent commented on src/SensioLabs/JobBoardBundle/Controller/BackendController.php:
```diff
@@ -19,11 +21,32 @@ public function listAction(Request $request)
+            $em->flush();
+
+            if ($job->getValidated()) {
+                $this->get('jobboard.mailer')->sendValidatedEmail($job, $oldTitle, $oldDescription);
```
I do not understand why an email about the validation is sent when a job is updated. It should be an email about the update of the job.

Jean-Beru replied:
From issue #8:
`When the announcement is validated by the administrator, an email is sent to the user to inform him that his announcement is now online, with a link to the announcement`

This email is sent only when it has been validated so I think it's more a "validation email" than an "update email".

jeremyFreeAgent replied:
Then the `$oldTitle` and `$oldDescription` are useless here since it is a validation email.
Plus the template use in the `sendValidatedEmail` should not be `updateNotification`.

jeremyFreeAgent replied:
The process is weird here. On each update of the job with this action the email of the validation will be sent even if the job was already validated.

Jean-Beru replied:
I think I misunderstood the issue.  This email should list all brought modifications. The "validation" checkbox should only pass status to "Published" or "New".

jeremyFreeAgent commented:
You should add an unit test for the mailer.

ogizanagi commented on src/SensioLabs/JobBoardBundle/Service/Mailer.php:
```diff
@@ -0,0 +1,49 @@
+                'oldDescription' => $oldDescription,
+            ]), 'text/html')
+        ;
+        $this->mailer->send($message);
```
No logs in case of errors ?

Jean-Beru commented:
@jeremyFreeAgent Mailer is already tested here : https://github.com/poledev/hubert/blob/issue-8/src/SensioLabs/JobBoardBundle/TestsFunctional/Controller/BackendControllerTest.php#L87

A unit testing will be like make the same thing again.

jeremyFreeAgent commented:
https://github.com/poledev/hubert/blob/issue-8/src/SensioLabs/JobBoardBundle/TestsFunctional/Controller/BackendControllerTest.php#L87 is not a unit test on the public method `sendValidatedEmail` of that mailer.

Jean-Beru commented:
@jeremyFreeAgent  OK, I see. So I added tests using mocks.
@ogizanagi You're right, I added `try... catch...` and logs in order to avoid exception pages.

salahm commented on src/SensioLabs/JobBoardBundle/Controller/BackendController.php:
```diff
@@ -19,11 +20,30 @@ public function listAction(Request $request)
+            $this->getDoctrine()->getManager()->flush();
+
+            if ($job->isPublished()) {
+                $this->get('jobboard.mailer')->sendUpdateEmail($job, $oldJob);
```
IMO this should be in a listener.

Jean-Beru replied:
Do you mean a form listener or an entity listener? The second one seems to be better.

ogizanagi replied:
Not a form listener.
A simple event dispatched using the symfony event dispatcher and an appropriated listener, or a Doctrine listener.
Both approaches are valid, @salahm would prefer a Doctrine listener, whereas I would rather dispatch my own event. The only difference is that with Doctrine, you're stuck if someday you have to not send the mail automatically when updating a job, unless adding a disabling mechanism (i.e a boolean field in the listener and opt-out in appropriated action). IMO, opt-in is better.

salahm replied:
:+1:

Jean-Beru replied:
OK, I see. I prefer Doctrine listeneners but you have more control with a simple dispatched event. So dispatched event wins!

salahm commented on src/SensioLabs/JobBoardBundle/Entity/Job.php:
```diff
@@ -357,11 +362,19 @@ public static function getStatuses()
+     */
+    public function isPublished()
+    {
+        return $this->status === self::STATUS_PUBLISHED;
```
please use yoda notation

salahm commented on src/SensioLabs/JobBoardBundle/Resources/views/Backend/list.html.twig:
```diff
@@ -55,7 +55,7 @@
                                     <td> 40 </td>
                                     <td>
-                                        <a href="{{ path('backend_edit') }}"> Edit </a>
+                                        <a href="{{ path('backend_edit', {id: 0}) }}"> Edit </a>
```
id 0 ?

Jean-Beru replied:
The list view didn't have any job ID at this moment. It's corrected with my recent rebase.

salahm commented on src/SensioLabs/JobBoardBundle/Service/Mailer.php:
```diff
@@ -0,0 +1,66 @@
+     * @param Job $job
+     * @param Job $oldJob
+     */
+    public function sendUpdateEmail(Job $job, Job $oldJob)
```
i don't think it's the resposability of this service to know which mail he's sending.
you should have only one method `send`, which will just "pass parameters to a view and send email"

ogizanagi commented:
:+1: once the listener issue is resolved.

Jean-Beru commented:
Listener issue solved

ogizanagi commented:
:+1:

ogizanagi commented:
You forgot to squash :cry:  !

Jean-Beru commented:
I didn't squash immediatly because I was waiting for my last commit validation :disappointed:
